\documentclass[compress]{beamer}
%\documentclass{beamer}

\usepackage{multicol}
\usetheme{Warsaw}
\useoutertheme[subsection=false]{miniframes}

%\mode<presentation>

\defbeamertemplate*{footline}{shadow theme}
{%
  \leavevmode%
  \hbox{\begin{beamercolorbox}[wd=.5\paperwidth,ht=2.5ex,dp=1.125ex,leftskip=.3cm plus1fil,rightskip=.3cm]{author in head/foot}%
    \usebeamerfont{author in head/foot}\insertframenumber\,/\,\inserttotalframenumber\hfill\insertshortauthor
  \end{beamercolorbox}%
  \begin{beamercolorbox}[wd=.5\paperwidth,ht=2.5ex,dp=1.125ex,leftskip=.3cm,rightskip=.3cm plus1fil]{title in head/foot}%
    \usebeamerfont{title in head/foot}\insertshorttitle%
  \end{beamercolorbox}}%
  \vskip0pt%
}


\title{Combinatorial Optimization for Fast Scaffolding}
\subtitle{Master Research Internship}
\author{Ivaylo PETROV}
\institute{University of Rennes 1}
%\logo{\includegraphics[width=0.15\textwidth]{./rennes1}}
\titlegraphic{\includegraphics[width=0.15\textwidth]{./rennes1}}
\date{\today}


\begin{document}

\begin{frame}
  \titlepage
\end{frame}

\section*{Outline} % (fold)
\label{sec:Outline}
  \begin{frame}
    \tableofcontents
  \end{frame}
% section Outline (end)

\section{Introduction} % (fold)
\label{sec:Introduction}

\subsection*{Dummy} % (fold)
\label{sub:Dummy}
% subsection Dummy (end)

  \begin{frame}
    \frametitle{What are we going to talk about today}
    \begin{multicols}{2}
      {\includegraphics[width=0.48\textwidth]{./presentation/dna.jpg}}
      \columnbreak
      \begin{itemize}
        \item DNA
        \item How we can study it?
        \item Why is it important to be studied?
      \end{itemize}
  \end{multicols}
  \end{frame}

  \begin{frame}
    \frametitle{How do we study DNA}
    \begin{itemize}
      \item The goal is to find the sequence of nucleotides that form the DNA
        molecule
      \item We obtain sequence of nucleotides from short parts of the DNA that
        are positioned at random places of the DNA molecule
      \item We try to put them together
    \end{itemize}
  \end{frame}

  \begin{frame}
    \frametitle{A picture of the expected result}
    {\includegraphics[width=0.95\textwidth]{./presentation/dna3.jpg}}
  \end{frame}
  
  \begin{frame}
    \frametitle{Reads and contig assembly}
    {\includegraphics[width=0.95\textwidth]{./presentation/contig_assembly.png}}
  \end{frame}
  

  \begin{frame}
    \frametitle{Why is it important to study the DNA}
    \begin{multicols}{2}
      {\includegraphics[width=0.52\textwidth]{./presentation/nicholas_volker.jpg}}
      \columnbreak
      \null \vfill
      \begin{itemize}
        \item For medical reasons
        \item For better understanding of biology and evolution of living organisms
        \item To alter or create new living organisms that will help us
      \end{itemize}
      \vfill \null
    \end{multicols}
  \end{frame}

  \begin{frame}
    \frametitle{Obtaining the genome sequence using reference genome}
    \begin{multicols}{2}
      {\includegraphics[width=0.95\columnwidth, height=0.85\textheight]
        {./presentation/sequencing_data_aligned_to_a_reference.jpg}}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Should have a reference to start with
          \item The choice of reference can affect the final result, which is not
            acceptable
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}

  \begin{frame}
    \frametitle{De novo genome assembly}
    \begin{multicols}{2}
      {\includegraphics[width=0.48\textwidth]{./assembly.jpg}}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Obtain the reads
          \item Build contigs from the reads
          \item Build scaffolds from the contigs and mate-pairs
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}

  \begin{frame}
    \frametitle{Mate-pairs}
    \begin{multicols}{2}
      \vfill
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        {\includegraphics[width=\columnwidth]{./presentation/mate-pairs.png}}
      \end{minipage}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Information about a pair of reads
          \item Give relative orientation
          \item Give relative distance in the genome
          \item Might contain deviations from the actual values or errors
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}
% section Introduction (end)

\section{Related work} % (fold)
\label{sec:Related work}

\subsection*{Dummy2} % (fold)
\label{sub:Dummy2}
% subsection Dummy2 (end)

  \begin{frame}
    \frametitle{Complexity of the problem}
    \begin{itemize}
      \item There might be errors in every phase of the process
      \item The problem is NP-hard
      \item Even only the orientation of the contigs is NP-hard
    \end{itemize}
  \end{frame}

  \begin{frame}
    \frametitle{Graphs used by newer methods}
    \begin{multicols}{2}
      \vfill
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        {\includegraphics[width=\columnwidth,
        height=0.5\textheight]{./presentation/graph_double_nodes_per_contig.png}}
      \end{minipage}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Older methods often used de Bruijn graphs, but they are not going to
            be presented now
          \item The nodes are either contigs or contig ends (resulting in two nodes
            for each contig)
          \item Edges are either mate-pairs or sets of mate-pairs that are very
            similar or sets of mate-pairs that just connect the same contigs
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}

  \begin{frame}
    \frametitle{Common problem formulation}
    \begin{multicols}{2}
      \vfill
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        {\includegraphics[width=\columnwidth,
        height=0.9\textheight]{./presentation/contig_graphs_orientation2.png}}
      \end{minipage}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Define some constraints
          \item Minimize the number of violated constraints
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}

  \begin{frame}
    \frametitle{Based only on heuristics}
    \begin{itemize}
      \item Basic idea
      \item SOAPdenovo, Greedy Path-Merging
      \item Disadvantages
    \end{itemize}
  \end{frame}

  \begin{frame}
    \frametitle{Trying to use exact methods}
    \begin{multicols}{2}
      \vfill
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        {\includegraphics[width=\columnwidth,
        height=0.5\textheight]{./presentation/contig_graph.jpg}}
      \end{minipage}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Basic idea
          \item SOPRA, MIR, Opera, GRASS
          \item Disadvantages
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}

  \begin{frame}
    \frametitle{An optimization for partitioning the graph}
    \begin{multicols}{2}
      \vfill
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        {\includegraphics[width=\columnwidth, height=0.6\textheight]{./presentation/problem_fragmentation.png}}
      \end{minipage}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Use contig that is not overpassed by an edge
          \item Not always easy
          \item If there are big enough contigs it is always possible
          \item MIR relies very much on this, using some heuristic make the
            graph partitionable
          \item This is also used by SOPRA, Opera and GRASS
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}
% section Related work (end)

\section{Additional steps} % (fold)
\label{sec:Additional steps}

\subsection*{Dummy3.0} % (fold)
\label{sub:Dummy3.0}
% subsection Dummy3.0 (end)

  \begin{frame}
    \frametitle{Data Filtering}
    \begin{multicols}{2}
      \vfill
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        {\includegraphics[width=\columnwidth,
        height=0.3\textheight]{./presentation/data_filtering_trimmed.png}}
      \end{minipage}
      \vfill
      \columnbreak
      \begin{minipage}[c][0.8\textheight][c]{\columnwidth}
        \begin{itemize}
          \item Big N50 and N90 values are good (they measure the average
            length)
          \item Small scaffolds count is good
          \item Data filtering can greatly improve the work of contig assemblers
            and scaffolders
        \end{itemize}
      \end{minipage}
    \end{multicols}
  \end{frame}

  %\begin{frame}
    %\frametitle{Finishing and Gap closure}
    %A SHORT DESCRIPTION
  %\end{frame}

% section Additional steps (end)

\section{Conclusion} % (fold)
\label{sec:Conclusion}

\subsection*{Dummy3} % (fold)
\label{sub:Dummy3}
% subsection Dummy3 (end)

  \begin{frame}
    \frametitle{Ideas for improvements}
    \begin{itemize}
      \item Exploit the count of times a contig is met
      \item Try to provide a number of optimal solutions
    \end{itemize}
  \end{frame}
% section Conclusion (end)

\section*{Questions and Answers} % (fold)
\label{sec:Questions and Answers}

\subsection*{Dummy3.3} % (fold)
\label{sub:Dummy3.3}
% subsection Dummy3.3 (end)

  \begin{frame}
    \frametitle{Questions and Answers}
    \begin{multicols}{2}
      \begin{minipage}[c][0.5\textheight][c]{\columnwidth}
        \centering
        Thank you very much for your attention!
      \end{minipage}
      \columnbreak
      {\includegraphics[width=0.95\columnwidth]{./presentation/questions-and-answers.jpg}}
      \vfill
    \end{multicols}
  \end{frame}
% section Questions and Answers (end)

%\note{Some note that can be separately compiled. It can be part of a slide as
%well}

%*****************************************************************%
\section{References} % (fold)
\label{sec:References}
\subsection*{Dummy4} % (fold)
\label{sub:Dummy4}
% subsection Dummy4 (end)

  \begin{frame}[allowframebreaks]
    \frametitle{References}
    \tiny

    \begin{thebibliography}{widest entry}
      \bibitem{pevzner-et-all-2001} Pevzner PA, Tang H, Waterman MS. 2001. An
        Eulerian path approach to DNA fragment assembly. Proc Natl Acad Sci 98:
        9748–9753.
      \bibitem{grass} Gritsenko, A. A., Nijkamp, J. F., Reinders, M. J. T.,
        and de Ridder, D. (2012). GRASS: a generic algorithm for scaffolding
        next-generation sequencing assemblies. Bioinformatics, 28(11), 1429–37.
      \bibitem{pevzner} Pevzner Pevzner PA, Tang H: Fragment assembly with
        double-barreled data. Bioinformatics 2001, 17(Suppl 1):S225-233.
      \bibitem{velvet-scaffolding} D.R. Zerbino, G.K. McEwen, E.H. Margulies, E.
        Birney, Pebble and rock band: heuristic resolution of repeats and
        scaffolding in the velvet short-read de novo assembler, PLoS One 4 (2009)
        e8407.
      \bibitem{SOAPdenovo} R. Li, H. Zhu, J. Ruan, W. Qian, X. Fang, Z. Shi, Y. Li,
        S. Li, G. Shan, K. Kristiansen, H. Yang, J. Wang, De novo assembly of
        human genomes with massively parallel short read sequencing, Genome Res. 20
        (2009) 265 - 272.
      \bibitem{SOPRA} Dayarian, A., Michael, T.P. and Sengupta, A.M. (2011) SOPRA:
        scaffolding algorithm for paired reads via statistical optimization, BMC
        Bioinformatics, 11, doi:10.1186/1471-2105-11-345.
      \bibitem{greedy-path-merging} Huson DH, Reinert K, Myers EW: The greedy
        path-merging algorithm for Contig Scaffolding. Journal of the Acm 2002,
        49(5):603-615.
      \bibitem{SSPACE} Boetzer,M. et al. (2011) Scaffolding pre-assembled contigs
        using SSPACE. Bioinformatics, 27, 578–579.
      \bibitem{MIR} Salmela, L., Mäkinen, V., Välimäki, N., Ylinen, J., Ukkonen,
        E. (2011). Fast scaffolding with small independent mixed integer programs.
        Bioinformatics, 27(23), 3259-3265.
      \bibitem{SCARPA} Donmez, N., Brudno, M. (2013). SCARPA: scaffolding reads
        with practical algorithms. Bioinformatics, 29(4), 428-434.
      \bibitem{Opera} Gao,S. et al. (2011) Opera: reconstructing optimal genomic
        scaffolds with high-throughput paired-end sequences. In: Bafna,V. and
        Sahinalp,S. (eds) Research in Computational Molecular Biology. Vol. 6577 of
        Lecture Notes in Computer Science. Springer, Berlin/Heidelberg, pp. 437–451.
      \bibitem{westbrook-tarjan} Westbrook,J. and Tarjan,R.E. (1992) Maintaining
        bridge-connected and biconnected components on-line. Algorithmica, 7,
        433–464.
      \bibitem{quake} Kelley, D. R., Schatz, M. C., Salzberg, S. L. (2010). Quake:
        quality-aware detection and correction of sequencing errors. Genome Biol,
        11(11), R116.
      \bibitem{filtering-solid} Sasson, A., Michael, T. P. (2010). Filtering error
        from SOLiD output. Bioinformatics, 26(6), 849-850.
      \bibitem{finIs} Gao, S., Bertrand, D., Nagarajan, N. (2012). FinIS: improved
        in silico finishing using an exact quadratic programming formulation. In
        Algorithms in Bioinformatics (pp. 314-325). Springer Berlin Heidelberg.
      \bibitem{MUMmer} Delcher, A.L., Phillippy, A., Carlton, J. and Salzberg, S.L.
        (2002) Fast algorithms for large-scale genome alignment and comparison,
        Nucleic Acids Research, 30, 2478– 2483, doi:10.1093/nar/30.11.2478
      \bibitem{NP-hard} Kececioglu, J. D. and Myers, E. W. (1995). Combinatorial algorithms
        for DNA sequence assembly. Algorithmica, 13, 7-51
    \end{thebibliography}
  \end{frame}
% section References (end)

\end{document}
